spec/requests/posts_spec.rb

```diff
@@ -0,0 +1,237 @@
+
+    it "assigns a new post to @post" do
+      get :new
+      expect(assigns(:post)).to be_a_new Post
+    end
+
+    context "when user is not logged in" do
+      it "redirects to login page" do
+        log_out
+        get :new
+        expect(response).to redirect_to login_path
+      end
+    end
+  end
+
+  describe "POST /create" do
+    let(:params) { {post: {body: "Hello world", title: "Hello"}} }
+
+    context "create a new post" do
+      it "create a new post" do
+        expect do
+          post :create, params: params
+        end.to change(Post, :count).by 1
+        expect(Post.last.body).to eq params.dig(:post, :body)
+        expect(response).to redirect_to user_posts_path(user)
+      end
+
+      it "create a new post failed" do
+        post :create, params: {post: {body: "", title: ""}}
+        expect(response).to render_template :new
+      end
+    end
+
+    context "when user is not logged in" do
+      it "redirects to login page" do
+        log_out
+        post :create, params: params
+        expect(response).to redirect_to login_path
+      end
+    end
+  end
+
+  describe "GET /show" do
+    let(:post) { user.posts.first }
+
+    it "returns http success" do
+      get :show, params: {id: post.id}
+      expect(response).to have_http_status(:success)
+    end
+
+    it "render the show template" do
+      get :show, params: {id: post.id}
+      expect(response).to render_template :show
+    end
+
+    it "assigns the requested post to @post" do
+      get :show, params: {id: post.id}
+      expect(assigns(:post)).to eq post
+    end
+
+    context "when post is not found" do
+      it "redirects to user's posts page" do
+        get :show, params: {id: 0}
+        expect(response).to redirect_to root_path
+        expect(flash[:danger]).to eq I18n.t("post_not_found")
+      end
+    end
+  end
+
+  describe "GET /edit" do
+    let(:post) { user.posts.first }
+
+    it "returns http success" do
+      get :edit, params: {
+        id: post.id
+      }
+      expect(response).to have_http_status(:success)
+    end
+
+    it "render the edit template" do
+      get :edit, params: {id: post.id}
+      expect(response).to render_template :edit
+    end
+
+    it "assigns the requested post to @post" do
+      get :edit, params: {id: post.id}
+      expect(assigns(:post)).to eq post
+    end
+
+    context "when user is not logged in" do
+      it "redirects to login page" do
+        log_out
+        get :edit, params: {id: post.id}
+        expect(response).to redirect_to login_path
+      end
+    end
+
+    context "when not correct user" do
+      it "redirects to user's posts page" do
+        get :edit, params: {id: Post.last.id}
+        expect(response).to redirect_to user_posts_path(user)
+      end
+    end
+  end
+
+  describe "PATCH /update" do
+    let(:post) { user.posts.first }
+    let(:params) { {id: post.id, post: {body: "Hello world", title: "Hello"}} }
+
+    context "update post" do
```

Comment:
Khi nào thì dùng context khi nào dùng describer khi nào dùng it nhỉ

Reply:
describe: nhóm các test case theo đối tượng

context: nhóm các test dựa trên điều kiện hoặc trạng thái cụ thể.

it: định nghĩa ra các test case